src/ahbicht/utility_functions.py

```diff
@@ -26,3 +30,26 @@ async def gather_if_necessary(results_and_awaitable_results: List[Union[Result,
             # we are sure obj is of type T
             result.append(obj)  # type:ignore[arg-type]
     return result
+
+
+def tree_copy(lru_cached_parsing_func: Callable[[str], Tree]):
+    """
+    A decorator that returns copy of the cached result from the lru_cached_parsing_func.
+    Rationale: We want to cache the tree for various expressions because this is definitely faster than re-parsing it.
+    But we don't want the same instance of the tree to be returned over and over again, because the calling code might
+    modify the tree and (if we always returned the same instance) might also modify the cache entry. We don't want this.
+    The tree_copy decorator is used together shall be used with a @lru_cached function. It returns a copy of the cached
+    value instead of the same instance.
+    :param lru_cached_parsing_func: A function that parses a string to a tree which is decorated with @lru_cache.
+    :return: the decorated function that always returns a copy of the cached result instead of the same instance
+    """
+
+    def decorated(*args, **kwargs):
+        cache_size_before_parsing = lru_cached_parsing_func.cache_info().currsize
+        tree_result: Tree = lru_cached_parsing_func(*args, **kwargs)
+        cache_size_after_parsing = lru_cached_parsing_func.cache_info().currsize
+        if cache_size_after_parsing == cache_size_before_parsing:
+            parsing_logger.debug("The parsed tree for '%s' has been loaded from the cache", args[0])
```

Comment:
Kann/Soll man keine f-strings im logger verwenden?

Reply:
ja, man soll keine string interpolation machen in log messages. Das gilt auch für C# (rider meckert das an) oder go.

Reply:
in pylint ist das auch ne eigene fehlermeldung. "fstring-in-log" oder so ähnlich: https://stackoverflow.com/questions/34619790/pylint-message-logging-format-interpolation

Reply:
Es geht also hauptsächlich um Performance oder?
Es wäre keine Sicherheitslücke wenn ich f-strings nutze oder?

Reply:
ne, es geht glaube ich darum, dass man leicht die stelle findet, an der eine bestimmte log-message geworfen wurde... und es ist wohl schneller... das erstbeste ergebnis auf google:
https://fixpython.com/2022/08/fix-python-pylint-message-logging-format-interpolation-45buo9k/